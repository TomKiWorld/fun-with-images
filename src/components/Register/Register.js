import React, { Component } from 'react';
import FormInput from '../FormFields/FormInput/FormInput';
import FormSubmit from '../FormFields/FormSubmit/FormSubmit';

/**
 * Register component
 * - Allows a user to register to the app by displaying a form
 * 
 * Required props:
 * - loadUser => Functon to get user information from state
 * - onRouteChange => Function to change the state of route
 */
class Register extends Component {
  constructor(props) {
    super();
    this.state = {
      name: '',
      email: '',
      password: '',
      nameError: '',
      emailError: '',
      passwordError: '',
      registerError: ''
    }
  }

  // Allow pressing the enter key and not only the button
  handleKeyUp = (e) => {
    if (e.keyCode === 13) {
      const register = document.querySelector('.register');
      register.click();
    }
  }

  // Sanitize user input
  sanitize = (string) => {
    const map = {
        '&': '&amp;',
        '<': '&lt;',
        '>': '&gt;',
        '"': '&quot;',
        "'": '&#x27;',
        '/': '&#x2F;',
    };
    const reg = /[&<>"'/]/ig;
    return string.replace(reg, (match)=>(map[match]));
  }

  // Set name on input change
  onNameChange = (event) => {
    this.setState({name: event.target.value })
  }

  // Set email on input change
  onEmailChange = (event) => {
    this.setState({email: event.target.value })
  }

  // Set password on input change
  onPasswordChange = (event) => {
    this.setState({password: event.target.value })
  }

  // Validate user input and register the user
  onSubmitRegister = (e) => {
    e.preventDefault();
    let {name, email, password } = this.state;
    let validName = false;
    let validEmail = false;
    let validPassword = false;
    const regexpEmail = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/;

    if (name !== '' ) {
      name = this.sanitize(name)
      this.setState({name: name})
      validName = true
      this.setState({nameError: ''});
    } else {
      this.setState({nameError: 'Please enter your name'});
    }

    if ((email !== '') && (regexpEmail.test(email))) {
      validEmail = true
      this.setState({emailError: ''});
    } else {
      this.setState({emailError: 'Please enter your email'});
    }

    if (password !== '' ) {
      validPassword = true
      this.setState({passwordError: ''});
    } else {
      this.setState({passwordError: 'Please enter your password'});
    }  

    if ( validName && validEmail && validPassword ) {
      fetch(`${this.props.database}/register`, {
        method: 'post',
        headers: {'Content-type': 'application/json'},
        body: JSON.stringify({
          name: name,
          email: email,
          password: password
        })
      })
      .then(response => response.json())
      .then(user => {
        if (user.id) {
          this.props.loadUser(user);
          this.props.onRouteChange('home');
        }
      })
      this.setState({registerError: ''});
    } else {
      this.setState({registerError: 'Something went wrong'});
    }
  }

  render() {
    const nameError = this.state.nameError ? <p className='error-message'>{this.state.nameError}</p> : '';
    const emailError = this.state.emailError ? <p className='error-message'>{this.state.emailError}</p> : '';
    const passwordError = this.state.passwordError ? <p className='error-message'>{this.state.passwordError}</p> : '';
    const registerError = this.state.registerError ? <p className='error-message'>{this.state.registerError}</p> : '';

    return (
      <article className='register br3 ba dark-gray b--black-10 mv4 mw6 shadow-5 center'>
        <form className='pa4 black-80'>
          <div className='measure'>
            <fieldset id='sign_up' className='ba b--transparent ph0 mh0'>
              <legend className='f1 fw6 ph0 mh0'>Register</legend>
              <FormInput 
                label='Name'
                name='name'
                type='text'
                onChange={this.onNameChange}
                onKeyUp={this.handleKeyUp}
              />
              {nameError}
              <FormInput 
                label='Email'
                name='email-address'
                type='email'
                onChange={this.onEmailChange}
                onKeyUp={this.handleKeyUp}
              />
              {emailError}
              <FormInput 
                label='Password'
                name='password'
                type='password'
                onChange={this.onPasswordChange}
                onKeyUp={this.handleKeyUp}
              />
              {passwordError}
            </fieldset>
            <FormSubmit
              value='Register'
              type='submit'
              onClick={this.onSubmitRegister}
              extraClass='register'
            />
            {registerError}
          </div>
        </form>        
      </article>
    );
  }
}

export default Register;
