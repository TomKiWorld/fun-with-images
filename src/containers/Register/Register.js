import React, { Component } from 'react';
import FormInput from '../../components/FormFields/FormInput/FormInput';
import FormSubmit from '../../components/FormFields/FormSubmit/FormSubmit';
import Preloader from '../../components/Preloader/Preloader';
import { DATABASE } from '../../constants';

/**
 * Register component
 * - Allows a user to register to the app by displaying a form
 * 
 * Required props:
 * - loadUser => Functon to get user information from state
 * - onRouteChange => Function to change the state of route
 */
class Register extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      password: '',
      nameError: '',
      emailError: '',
      passwordError: '',
      registerError: '',
      loading: false
    }
  }

  // Sanitize user input
  sanitize = (string) => {
    const map = {
        '&': '&amp;',
        '<': '&lt;',
        '>': '&gt;',
        '"': '&quot;',
        "'": '&#x27;',
        '/': '&#x2F;',
    };
    const reg = /[&<>"'/]/ig;
    return string.replace(reg, (match)=>(map[match]));
  }

  // Set state on input change
  onFieldChange = (event, term) => {
    this.setState({[term]: event.target.value })
  }

  // Validate user input and register the user
  onSubmitRegister = (e) => {
    e.preventDefault();
    let {name, email, password } = this.state;
    let validName = false;
    let validEmail = false;
    let validPassword = false;
    const regexpEmail = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/;

    if (name !== '' ) {
      name = this.sanitize(name)
      this.setState({name: name})
      validName = true
      this.setState({nameError: ''});
    } else {
      this.setState({nameError: 'Please enter your name'});
    }

    if ((email !== '') && (regexpEmail.test(email))) {
      validEmail = true
      this.setState({emailError: ''});
    } else {
      this.setState({emailError: 'Please enter your email'});
    }

    if (password !== '' ) {
      validPassword = true
      this.setState({passwordError: ''});
    } else {
      this.setState({passwordError: 'Please enter your password'});
    }  

    if ( validName && validEmail && validPassword ) {
      this.setState({loading: true})
      fetch(`${DATABASE}/register`, {
        method: 'post',
        headers: {'Content-type': 'application/json'},
        body: JSON.stringify({
          name: name,
          email: email,
          password: password
        })
      })
      .then(response => response.json())
      .then(user => {
        if (user.id) {
          this.props.loadUser(user);
          this.setState({
            registerError: '',
            loading: false
          });
          this.props.onRouteChange('home');
        } else {
          this.setState({
            registerError: user,
            loading: false
          });
        }
      })
      .catch(err => {
        this.setState({
          registerError: 'Something went wrong please try again later',
          loading: false
        });
      })
    } else {
      this.setState({registerError: 'Something went wrong'});
    }
  }

  render() {
    const nameError = this.state.nameError ? <p className='error-message'>{this.state.nameError}</p> : '';
    const emailError = this.state.emailError ? <p className='error-message'>{this.state.emailError}</p> : '';
    const passwordError = this.state.passwordError ? <p className='error-message'>{this.state.passwordError}</p> : '';
    const registerError = this.state.registerError ? <p className='error-message'>{this.state.registerError}</p> : '';
    const preload = this.state.loading ? <Preloader/> : '';

    return (
      <article className='register br3 ba dark-gray b--black-10 mv4 mw6 shadow-5 center'>
        <form className='pa4 black-80'>
          <div className='measure'>
            <fieldset id='sign_up' className='ba b--transparent ph0 mh0'>
              <h1 className='f1 fw6 ph0 mh0'>Register</h1>
              <FormInput 
                label='Name'
                name='name'
                type='text'
                onChange={(e) => this.onFieldChange(e, 'name')}
              />
              {nameError}
              <FormInput 
                label='Email'
                name='email-address'
                type='email'
                onChange={(e) => this.onFieldChange(e, 'email')}
              />
              {emailError}
              <FormInput 
                label='Password'
                name='password'
                type='password'
                onChange={(e) => this.onFieldChange(e, 'password')}
              />
              {passwordError}
            </fieldset>
            <FormSubmit
              value='Register'
              type='submit'
              onClick={this.onSubmitRegister}
              extraClass='register'
            />
            {registerError}
          </div>
        </form>
        {preload}     
      </article>
    );
  }
}

export default Register;
