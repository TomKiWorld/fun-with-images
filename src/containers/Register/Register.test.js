import { shallow } from 'enzyme';
import React from 'react';
import Register from './Register';

describe('Register Component', () => {
  const wrapper = shallow(<Register />);
  const nameInput = wrapper.find('[name=\'name\']').at(0);
  const emailInput = wrapper.find('[name=\'email-address\']').at(0);
  const passwordInput = wrapper.find('[name=\'password\']').at(0);
  
  it('Expect to render Register Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Expect to sanitize string', () => {
    expect.assertions(7);
    expect(wrapper.instance().sanitize('&')).toBe('&amp;');
    expect(wrapper.instance().sanitize('<')).toBe('&lt;');
    expect(wrapper.instance().sanitize('>')).toBe('&gt;');
    expect(wrapper.instance().sanitize('"')).toBe('&quot;');
    expect(wrapper.instance().sanitize("'")).toBe('&#x27;');
    expect(wrapper.instance().sanitize('/')).toBe('&#x2F;');
    expect(wrapper.instance().sanitize('test')).toBe('test');
  });

  it('Expect to handle error submit register', () => {
    const submitBtn = wrapper.find('[value=\'Register\']').at(0);
    submitBtn.simulate('click', { preventDefault: jest.fn() });
    expect.assertions(5);
    expect(wrapper.state().nameError).toEqual('Please enter your name');
    expect(wrapper.state().emailError).toEqual('Please enter your email');
    expect(wrapper.state().passwordError).toEqual('Please enter your password');
    expect(wrapper.state().registerError).toEqual('Something went wrong');
    expect(wrapper.state().loading).toEqual(false);
  });

  it('Expect to handle name change', () => {
    nameInput.simulate('change', { target: { value: 'John', term: 'name'} });
    expect(wrapper.state().name).toEqual('John');
  });

  it('Expect to handle email change', () => {
    emailInput.simulate('change', { target: { value: 'john@gmail.com', term: 'email'} });
    expect(wrapper.state().email).toEqual('john@gmail.com');
  });

  it('Expect to handle password change', () => {
    passwordInput.simulate('change', { target: { value: 'unsafePassword', term: 'password'} });
    expect(wrapper.state().password).toEqual('unsafePassword');
  });

  it('Expect to handle submit register', () => {
    const submitBtn = wrapper.find('[value=\'Register\']').at(0);
    submitBtn.simulate('click', { preventDefault: jest.fn() });
    expect(wrapper.state().loading).toEqual(true);
  });
});
