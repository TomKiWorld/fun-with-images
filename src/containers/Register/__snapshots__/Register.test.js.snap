// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component Expect to render Register Component 1`] = `ShallowWrapper {}`;
